### PR TITLE
prov/rxm: minor refactoring + bug fix and updates for fabtests

### PR DESCRIPTION
**Pull request by a-ilango**
- prov/rxm: add some logging 
- fabtests: use size_t for some size types to fix comparison warnings 
- fabtests: add data integrity verification to multi recv test 
- prov/rxm: merge op_flags and flags args in rxm_ep_recv_common_flags 
- prov/rxm: remove RXM_INJECT_TX state from MSG CQ completion handler

### CODE REVIEW

**Comment by a-ilango** [comment]
AWS CI sockets provider error seems like a timing issue of when server and client are started;

```
13:47:30 /home/ec2-user/workspace/multi-node/PROVIDER/sockets/label/rhel/fabtests/install/bin/runfabtests.sh: line 379: 27259 Killed                  ${SERVER_CMD} "${EXPORT_ENV} $s_cmd" &>$s_outp
13:47:30 - name:   fi_rma_bw -e msg -o writedata -I 5 -p "sockets"
13:47:30   result: Fail
13:47:30   time:   1
13:47:30   server_cmd: /home/ec2-user/workspace/multi-node/PROVIDER/sockets/label/rhel/fabtests/install/bin/fi_rma_bw -e msg -o writedata -I 5 -p "sockets" -s 172.31.22.224
13:47:30   server_stdout: |
13:47:30   client_cmd: /home/ec2-user/workspace/multi-node/PROVIDER/sockets/label/rhel/fabtests/install/bin/fi_rma_bw -e msg -o writedata -I 5 -p "sockets" -s 172.31.24.198 172.31.22.224
13:47:30   client_stdout: |
13:47:30     fi_connect(): common/shared.c:821, ret=-111 (Connection refused)
```

File: prov/rxm/src/rxm_cq.c [hunk 1]
```diff
@@ -1217,15 +1217,19 @@ static void rxm_cq_read_write_error(struct rxm_ep *rxm_ep)
...
-		    rx_buf->hdr.desc, FI_ADDR_UNSPEC, rx_buf)) {
-		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL, "Unable to repost buf\n");
-		return -FI_EAVAIL;
+	ret = fi_recv(rx_buf->msg_ep, &rx_buf->pkt,
+		      rx_buf->ep->eager_limit + sizeof(struct rxm_pkt),
+		      rx_buf->hdr.desc, FI_ADDR_UNSPEC, rx_buf);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to repost buf: %d\n", ret);
+		return ret;
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: nit: move 'return ret' to end of function - replace 'return FI_SUCCESS'

File: prov/rxm/src/rxm_ep.c [hunk 2]
```diff
@@ -744,9 +745,19 @@ rxm_ep_post_recv(struct rxm_ep *rxm_ep, const struct iovec *iov,
...
+	if (recv_queue->type == RXM_RECV_QUEUE_MSG)
+		FI_DBG(&rxm_prov, FI_LOG_EP_DATA, "Posting recv with length: %zu "
+		       "addr: 0x%" PRIx64 "\n", recv_entry->total_len,
+		       recv_entry->addr);
+	else
+		FI_DBG(&rxm_prov, FI_LOG_EP_DATA, "Posting trecv with "
+		       "length: %zu addr: 0x%" PRIx64 " tag: 0x%" PRIx64
+		       " ignore: 0x%" PRIx64 "\n", recv_entry->total_len,
+		       recv_entry->addr, recv_entry->tag, recv_entry->ignore);
+#endif
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c]: You may be able to drop the ENABLE_DEBUG.  The if statements will be empty and should be compiled out anyway.

File: prov/rxm/src/rxm_cq.c [hunk 3]
```diff
@@ -999,12 +1000,16 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
...
-	case RXM_INJECT_TX:
 		tx_eager_buf = comp->op_context;
 		assert(comp->flags & FI_SEND);
 		ret = rxm_finish_eager_send(rxm_ep, tx_eager_buf);
 		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX, tx_eager_buf);
 		return ret;
+	case RXM_INJECT_TX:
+		tx_base_buf = comp->op_context;
+		assert(comp->flags & FI_SEND);
+		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX_INJECT, tx_base_buf);
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: What is the full impact of this bug?  Do the buffer pools maintain similar structures, such that one can be re-used, and we're really only looking at a memory leak?  Or is this something worse?

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: So, one of the differences between the two pools is that buffers from RXM_BUF_POOL_TX_INJECT wouldn't have hdr field set. I'm surprised we haven't seen a segfault when functions using buffers from RXM_BUF_POOL_TX try to access the hdr.

File: prov/rxm/src/rxm_ep.c (hunk 2 above)
Comment (a-ilango) [reply prov/rxm/src/rxm_ep.c]: Good point. Will update.

File: prov/rxm/src/rxm_cq.c (hunk 3 above)
Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: I *just* fixed a similar problem in the tcp provider, though the likelihood of hitting that issue would have been rare.  This is obviously something that we need to try to detect. It looks like we should have this ability already with a debug build, so I'll try to add an appropriate assert.

We should try to figure out why we aren't seeing constant crashes without this change.

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: It looks like we don't need this fix at all. Buffers from RXM_BUF_POOL_TX_INJECT are used only for fi_inject calls and that would never generate a completion from MSG provider. So we would never hit this switch case.

Also, the buffers are returned back to the right pool after fi_inject succeeds. So there wouldn't any sort of memory leak.

I'll drop the patch.

Comment (shefty) [reply prov/rxm/src/rxm_cq.c]: Does that mean the location of 'case RXM_INJECT_TX' is correct, or that we shouldn't actually hit it?  Should this fall through to the default case or have an assertion?

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: We shouldn't hit the case. I'll add an assertion that the state is not RXM_INJECT_TX when we get a completion.

Also, it seems almost all of the fabtests set FI_THREAD_DOMAIN. We would use buffers from the inject pool only when it's FI_THREAD_SAFE. We're missing some test coverage there. I think fi_ubertest doesn't have a threading level config as well.

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: fi_ubertest should test FI_THREAD_SAFE as that's the default threading level would be returned. I dropped this commit and added a new one that adds the assert.

File: prov/rxm/src/rxm_cq.c [hunk 4]
```diff
@@ -1217,17 +1218,19 @@ static void rxm_cq_read_write_error(struct rxm_ep *rxm_ep)
...
-		return -FI_EAVAIL;
-	}
-	return FI_SUCCESS;
+	ret = fi_recv(rx_buf->msg_ep, &rx_buf->pkt,
+		      rx_buf->ep->eager_limit + sizeof(struct rxm_pkt),
+		      rx_buf->hdr.desc, FI_ADDR_UNSPEC, rx_buf);
+	if (ret)
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to repost buf: %d\n", ret);
+	return ret;
```
Comment (shefty) [inline prov/rxm/src/rxm_cq.c]: This function is converting ssize_t (from fi_recv) into an int (ret).  I don't know if you want to convert rxm_ep_repost_buf from int to ssize_t, or just apply a cast.  If you go with the first option, then series is okay to merge, with a follow up cleanup patch at some point.

Comment (a-ilango) [reply prov/rxm/src/rxm_cq.c]: Sure. I'll convert int to ssize_t in a follow-up patch.

**Pull request merged**